app/models/vm_or_template.rb

```diff
@@ -194,6 +194,14 @@ class VmOrTemplate < ApplicationRecord
   scope :orphaned,  ->       { where(:ems_id => nil).where.not(:storage_id => nil) }
   scope :with_ems,  ->       { where.not(:ems_id => nil) }
 
+  # The SQL form of `#registered?`, with it's inverse as well.
+  # TODO: Vmware Specific (copied (old) TODO from #registered?)
+  scope :registered, (lambda do
+    where(arel_table[:template].eq(false).or(arel_table[:ems_id].not_eq(nil)).and(arel_table[:host_id].not_eq(nil)))
```

Comment:
gah, needed to do de-morgan, since my brain didn't take it's correct :-D

 ! ((template && !ems_id) or !host_id)
that is
!(template && !ems_id) && host_id
and
(!template || ems_id) && host_id

Reply:
Yeah, I needed the tests in there because otherwise I was not going to get it right.

Reply:
Also, De-morgan was what I was thinking of, and all those other rules I have long since forgotten...

Reply:
I have derived ruby logic like this too many times.  Just had to do this for  #16577
```ruby
!(self.tamplate? && ext_management_system_id.nil?) && host_id.present?
(!template? || !ext_management_sytem_id.nil?) && !host_id.nil?
```
Wonder if we have `nil` in this column :(

Will this work with AR instead of arel?

```ruby
where(:template => false).or(where.not(:ems_id => nil)).where.not(:host_id => nil)
```

Reply:
> Will this work with AR instead of arel?

Possibly.  I can run it through the tests and see (a big reason I wrote them).

> Wonder if we have `nil` in this column :(

what column are you referring to?  There are three. (also, `s/tamplate/template/`)

Reply:
> um spelling...

thnx

> wonder if `template` can be `nil` in the database.

answer: not a concern

```
    t.boolean  "template",                 default: false
```

Reply:
With default false, you can still set a nil right? There is no ```not null``` constraint. :-)